src/components/contents-context.js

```diff
@@ -55,62 +54,64 @@ const findSectionHeadings = (nodes) => {
   return results;
 };
 
-const findHeadingsOnPage = (headingNodes, height) => {
-  let positions = [];
-
+const observeHeadings = (headingNodes, observer) => {
+  let headingElements = [];
   headingNodes.forEach((heading) => {
     const el = document.getElementById(heading.id);
     if (el) {
-      positions.push(el.offsetTop / height);
+      observer.observe(el);
+      headingElements.push(el);
     }
   });
+  return headingElements;
+};
 
-  return positions;
+const unobserveHeadings = (headings, observer) => {
+  headings.forEach((el) => {
+    observer.unobserve(el);
+  });
 };
 
 const ContentsContext = React.createContext(defaultContextValue);
 
 const ContentsProvider = ({ children, nodes = [] }) => {
-  const [activeSectionIndex, setActiveSectionIndex] = useState(0);
-  const [headingNodes, setHeadingNodes] = useState([]);
+  const headingNodes = useMemo(() => findSectionHeadings(nodes), [nodes]);
+  const [activeSectionId, setActiveSectionId] = useState(headingNodes[0]?.id);
 
   useEffect(() => {
-    setHeadingNodes(findSectionHeadings(nodes));
-  }, [nodes]);
+    const options = {
+      root: document.querySelector('.content'),
+      rootMargin: '0px',
+      threshold: 1.0,
+    };
 
-  useEffect(() => {
-    const height = document.body.clientHeight - window.innerHeight;
-    const headingPositions = findHeadingsOnPage(headingNodes, height);
-
-    const handleScroll = () => {
-      // Calculate current position of the page similar to guides.js
-      const scrollTop = Math.max(window.pageYOffset, document.documentElement.scrollTop, document.body.scrollTop);
-      let currentPosition = scrollTop / height;
-      currentPosition = (scrollTop + currentPosition * 0.8 * window.innerHeight) / height;
-
-      // bestMatch = [distance from closest section, closest section index]
-      let bestMatch = [Infinity, 0];
-
-      headingPositions.forEach((headingPosition, index) => {
-        const delta = Math.abs(headingPosition - currentPosition);
-        if (delta <= bestMatch[0]) {
-          bestMatch = [delta, index];
+    // Callback is first performed upon page load. Ignore checking entries on first load to allow
+    // headings[0] to be counted first when headings[0] and headings[1] are both intersecting
+    let firstLoad = true;
+    let defaultActiveSectionId = window.location.hash.slice(1) || headingNodes[0]?.id;
```

Comment:
[nit] The ultimate safest way to do this would be:
```suggestion
    let defaultActiveSectionId = window.location.hash.slice(1) || headingNodes?.[0]?.id;
```

Reply:
Despite being the ultimate safest way to do this, I wonder if this is necessary. `headingNodes` should be expected to exist because we get it from [`findHeadings`](https://github.com/mongodb/snooty/pull/363/files/670c9174bbae7a54ee92ee5c2b37213e51f859d6#diff-b45effd2f2fad9df3bc7f776ee52abfa2b52a8f9a73ea8648d4ba355fd383a7fR13), which at the very least should return an empty array.